Add `types.attrTag`

roberth opened the pull request:
## Description of changes

The purpose of `attrTag` is to describe values that are already structured with this exact syntax, in RFC 42 settings.

It can also be used as an alternative to other sum type representations that aren't merged yet (edit: in some cases). This may or may not be nice, but that's irrelevant because `attrTag` is needed for a specific purpose anyway, as mentioned.
My goal here is only to introduce _mechanism_, not _policy_. For instance, if we want to pick a favorite sum type representation, or formulate some other kinds of recommendations, that can be done separately, later, when we have practical experience with multiple sum types.

I've updated the docs to lay out the new concept and how it relates to the existing "union-like" types.

Other sum type in development (not suitable for my use case as mentioned)
- #254790

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
lib.types.attrTag: init

roberth committed:
doc/option-types: Write about sum and union types

roberth committed:
doc/option-types: Move either and oneOf to Union types

roberth committed:
doc/option-types: Move attrTag to Sum types, remove redundant paragraph

roberth committed:
doc/option-types: Move nullOr to Union types

roberth committed:
lib.types.attrTag: Support module docs

roberth committed:
lib.types.attrTag: Take options instead of types

roberth committed:
lib.types.attrTag: Support type merging

roberth committed:
lib/types.nix: Fix getSubOptions doc

This comment was added in 73f32d03758a5, when it was already
supposed to be an attrset.

roberth committed:
lib.types.attrTag: Provide declarations, definitions

roberth committed:
lib.modules.evalOptionValue: Undeprecate for lib.types

lib.types.attrTag needs it.

roberth committed:
lib.types.attrTag: Remove tags from description

You can find them in the sub-options now.

roberth committed:
doc/option-types: Add attrTag example

infinisil commented on lib/types.nix:
```diff
@@ -613,6 +618,95 @@ rec {
+        nestedTypes = tags;
+        functor = (defaultFunctor "attrTagWith") // {
+          payload = { inherit tags; };
+          binOp =
```
There's a potential minor issue with supporting type merging, because users may write code like this:

```nix
{
  tag1 = 0;
  tag2 = 1;
}.${config.my.option}
```

And if some third-party module specifies another tag, this code breaks, without any good way to fix it.

Though I guess if we document that code needs to handle arbitrarily many tags, it might be fine.

roberth replied:
Your example looks like a usage of `enum`.
Indeed `attrTag` has that same characteristic, but I don't think this will happen by accident.

Regarding the sameness:
- `enum [str1 .. strn]` is dual to an attrset of values.
- `attrTag ...` is dual to an attrset of functions.

infinisil commented on lib/types.nix:
```diff
@@ -613,6 +618,95 @@ rec {
+                };
+            })
+            tags;
+        substSubModules = m: attrTagWith { tags = mapAttrs (n: opt: opt // { type = (opt.type or types.unspecified).substSubModules m; }) tags; };
```
The `types.unspecified` looks odd here, is that necessary or can we avoid it?

roberth replied:
In a module, the presence of `opt.type` would be guaranteed by `fixupOptionType`.
Calling that seems heavy-handed considering that this works, and is basically free, in terms of eval performance.

roberth replied:
I've removed `substSubModules` because without `getSubmodules`, this turns out to be dead code anyway.
So I've found while trying to test it.

I don't think `getSubmodules` is feasible to implement, because we don't have a single set of modules to return, let alone put back with `substSubModules`

I'll be happy to do something about this later if we find some sub-optimal generated docs.

infinisil commented on lib/types.nix:
```diff
@@ -613,6 +618,95 @@ rec {
+      in
+      mkOptionType {
+        name = "attrTag";
+        description = "attribute-tagged union";
```
`descriptionClass` missing

infinisil reviewed (commented):
This is looking great! This really feels like a hole in the module system that nobody bothered filling until now.

There appears to be a problem when trying to build the manual with options of this type:

```nix
# configuration.nix
{ lib, ... }: {
  options.mine = lib.mkOption {
    type = lib.types.attrTag {
      a = lib.types.int;
    };
    description = "mine";
  };

  config.documentation.nixos.includeAllModules = true;
}
```

```
$ nix-build nixos --arg configuration ./configuration.nix -A config.system.build.manual.manual

       error: assertion '((opt)._type == "option")' failed

       at /home/tweagysil/src/nixpkgs/main/lib/types.nix:629:17:

          628|               builtins.addErrorContext "while checking that attrTag tag ${lib.strings.escapeNixIdentifier n} is an option with a type${inAttrPosSuffix args.tags n}" (
          629|                 assert opt._type == "option";
             |                 ^
          630|                 opt // {
```

lheckemann commented on nixos/doc/manual/development/option-types.section.md:
````diff
@@ -279,6 +282,67 @@ Submodules are detailed in [Submodule](#section-option-types-submodule).
+        type = submodule {};
+      };
+    }
+    ```
````
An example of possible values for an option of this type would be helpful.

My best guess would be that it allows
```nix
{ bounce.errorMessage = "foo bar"; }
```
and
```
{ ignore = {}; }
```
?

lheckemann commented on lib/tests/modules/types-attrTag.nix:
```diff
@@ -0,0 +1,113 @@
+    merged.extensi-foo.extensible = "foo";
+    merged.extensi-bar.extensible = "bar";
+    okChecks =
+      assert config.intStrings.hello.right == "hello world";
```
Should these maybe assert over the whole value of the attrset, i.e. `assert config.intStrings.hello == { right = "hello world"; }`?

roberth replied:
Good point! Done.

fricklerhandwerk commented on nixos/doc/manual/development/option-types.section.md:
````diff
@@ -279,6 +282,67 @@ Submodules are detailed in [Submodule](#section-option-types-submodule).
+    Example:
+
+    ```nix
+    types.attrTag {
````
Would be good to have a full (self-contained module) example including assigning `config` to use the declared option with one of the types.

roberth replied:
Got a bit lengthy, but worth it.

roberth committed:
doc/option-types: Make attrTag example self-contained

... well, except for the ellipses, which hide unnecessary descriptions,
which you should write!

roberth committed:
lib/tests/modules/types-attrTag: Test against unexpected attrs

Thank you lheckemann for pointing this out!

roberth committed:
lib.types.attrTag: Custom error when passing bare type

roberth commented:
> There appears to be a problem when trying to build the manual with options of this type:

> ```nix
>     type = lib.types.attrTag {
>       a = lib.types.int;
>     };
> ```

I've changed it to take options instead of types. This showed that `assert` was not good enough. I've added a custom error message for this just now.

Ma27 commented:
> It can also be used as an alternative to other sum type representations that aren't merged yet.

Just for the reference: this is not a substitute for all use-cases of `taggedSubmodule` though: for instance, in Nextcloud you have e.g. the option `dbtype` and if that's `sqlite`, `dbuser`/`dbpass`/etc are invalid whereas it is required for `dbtype` being `mysql` and `psql`.

infinisil commented:
@Ma27 I replied to your comment in the other PR since it's more relevant there: https://github.com/NixOS/nixpkgs/pull/254790#issuecomment-1965040917

Lassulus commented:
What is dislike about this solution, is that there is no apparent indicator that a specific value is a sum-type. it just looks like a normal attribute set and you either have to discover it through configuring it wrong or reading the options source code? with the taggedSubmodule type that is IMHO clearer, but this code is clearer.

roberth commented:
@Lassulus We need both types, because both occur in the wild, in RFC 42 settings.

> Other sum type in development

Not intended to compete, but a statement of fact. In a non-RFC 42 context they are alternatives to each other, and I don't think I care which one gets chosen by module authors.

infinisil commented:
> What is dislike about this solution, is that there is no apparent indicator that a specific value is a sum-type. it just looks like a normal attribute set and you either have to discover it through configuring it wrong or reading the options source code? with the taggedSubmodule type that is IMHO clearer, but this code is clearer.

That's fair. Maybe it could be this instead:

```
{
  type = "foo";
  foo = {
    # ...
  };
}
```

This would be a combination of `attrTag` and `taggedSubmodule`.

roberth commented:
> Maybe it could be this instead:
> 
> ```nix
> {
>   type = "foo";
>   foo = {
>     # ...
>   };
> }
> ```

To implement this, existing constructs and laziness may be enough, if you're not doing RFC 42 settings (`config` would have noise and/or be partial).

Example:

```nix
{ config, options, ... }: {
  options.widget = {
    type = mkOption { type = enum [ "foo" "bar" ]; };
    # Ignore when `type != "foo"`
    foo = mkOption { type = submodule fooModule; };
    # Or you could force it. Strictness can make infinite recursions happen, but seems unlikely.
    bar = mkOption {
      type = submodule barModule;
      apply = throwIf
        (options.widget.type.value != "bar")
        "${options.widget.bar} is not defined because ${options.widget.type} is ${lib.strings.escapeNixString options.widget.type.value} instead of \"bar\". ";
    };
  };
}
```

It is somewhat more extensible. If it turns out that the option has sub-options that are valid for all tags, you can just add them directly under `widget`.

Yet another possible syntax is `{ type = ...; value = ...; }`, which is also close to `attrTag`, through the isomorphism that uses `lib.nameValuePair`.

So we have these possible syntaxes. I've taken the liberty to name them (but better suggestions welcome).

- `attrTag`: `{ ${tag} = v; }`
- `typeTag`: `{ type = tag; value = v; }`
- `explicitTag`: `{ type = tag; ${tag} = v; }`
- `taggedSubmodule` `v // { type = tag; }` roughly

roberth commented:
> That's fair. Maybe it could be this instead:

My use case needs the "attrTag" syntax though. This thread is a little distracted; can we focus on the chosen syntax instead? Changing it is not a possibility for me, because it's for a service-specific type.

Lassulus commented:
Since we seem to plan to implement multiple sum-types. Does it make sense to categorize them somehow beforehand? Like having them in something like lib.types.sumTypes.attrTag?
Otherwise the changes look good to me and I would build upon them by implementing the next sum-type if this pr is merged.

nyabinary commented:
> > like lib.types.sumTypes.attrTag?
> 
> I'd be ok with either. @infinisil wdyt?

bump @infinisil

infinisil reviewed (changes requested):
Overall this is looking pretty good!

I think what's missing is a convenient way to use the resulting values. This works, but has a lot of boilerplate:

```nix
value:
let
  tag = lib.head (lib.attrNames value);
  tagValue = value.${tag};
in
{
  tag1 = "Tag1: ${tagValue}";
  tag2 = "Tag2: ${tagValue}";
}.${tag} or (throw "nope")
```

It seems sensible to introduce something like this:

```nix
value:
lib.matchAttrTag value fallback {
  tag1 = tagValue: "Tag1: ${tagValue}";
  tag2 = tagValue: "Tag2: ${tagValue}";
}
```

---

A minor problem I discovered is that options within submodules within tags don't have declaration position information, here's a reproducer:

```nix
# test.nix
with import ./lib;

let
  result = evalModules {
    modules = [
      ({ lib, ... }: {
        _file = "test.nix";
        options.tagged = lib.mkOption {
          description = "tagged";
          type = lib.types.attrTag {
            aTag = lib.mkOption {
              description = "atag";
              type = lib.types.submodule {
                options.foo = lib.mkOption {
                  description = "foo";
                  type = lib.types.int;
                };
              };
            };
          };
        };
      })
    ];
  };
in
optionAttrSetToDocList result.options
```

```
$ nix-instantiate --eval test.nix --strict --json | jq '.[] | select(.loc == [ "tagged", "aTag", "foo" ])'
{
  "declarations": [],
  "description": "foo",
  "internal": false,
  "loc": [
    "tagged",
    "aTag",
    "foo"
  ],
  "name": "tagged.aTag.foo",
  "readOnly": false,
  "type": "signed integer",
  "visible": true
}
```

roberth commented on lib/types.nix:
```diff
@@ -614,6 +620,102 @@ rec {
+                  );
+                }
+              ))
+            args.tags;
```
I've shadowed the original `tags` binding from the argument to make sure we use the checked and normalized one below.

infinisil replied:
I generally prefer an about-equally-ugly-but-more-explicit `let _tags = tags; in let tags = f _tags;`, but this is fine too

roberth replied:
Removing `attrTagWith` forced me to do it like this.

I believe that approach is slightly more expensive because of an extra `Env`, but this isn't hot code, so I'll just stop talking now :)

roberth commented on nixos/doc/manual/development/option-types.section.md:
```diff
@@ -279,6 +282,85 @@ Submodules are detailed in [Submodule](#section-option-types-submodule).
+
+## Sum types {#sec-option-types-sums}
+
+A sum type can be thought of, conceptually, as a *`types.enum`* where each valid item is paired with a type, through some value syntax.
+Nix does not have a built-in syntax for this pairing of a label and a type or value, so sum types may be represented in multiple ways.
+
+If the built-in Nix value types provide enough distinction, you simplify your syntax with a [union type](#sec-option-types-unions) instead.
+
+`types.attrTag` *`{ attr1 = t1; attr2 = t2; ... }`*
+
+:   An attribute set containing one attribute, whose name must be picked from
+    the attribute set (`attr1`, etc) and whose value must be of the accompanying
+    type.
```
Maybe something can be done to allow a static default of sorts, bypassing the `config` up to the point where the tag is decided. This would have to use https://github.com/NixOS/nixpkgs/pull/296979, because asking users to provide a `mkDefault` definition is not reasonable.
Might be something for a follow-up. Definitely needs docs.

roberth replied:
Done in https://github.com/NixOS/nixpkgs/pull/284551/commits/f354686536823a0bedb95abf13b090b961485bf9

I did write "at least a type" instead of "option", so that the introduction about sum _types_ remains general.

infinisil commented on lib/types.nix:
```diff
@@ -614,6 +620,102 @@ rec {
 
+    attrTag = tags: attrTagWith { inherit tags; };
+
+    attrTagWith = args@{ tags }:
```
@roberth I see you're [experimenting with more arguments](https://github.com/NixOS/nixpkgs/pull/284551/commits/45cafb71e37e74e82ab14ca82d46d085315c9ec2), but I'd prefer not adding more features to this PR. It was already big enough and effectively ready

roberth committed:
lib/tests/modules/types-attrTag.nix: Clean up unneeded comment

roberth committed:
types.attrTagWith: remove

Keep it simple for now.

roberth committed:
types.attrTag: Remove substSubmodules

I haven't managed to trigger the error, and it turns out that this
method is optional.
Specifically, getSubmodules is unimplemented (and unimplementable),
the tests pass, and we seem to have good location info.

roberth committed:
doc/option-types: Update sum types

roberth commented on lib/tests/modules/types-attrTag.nix:
```diff
@@ -0,0 +1,113 @@
+      assert config.merged.positive == { yay = 100; };
+      assert config.merged.extensi-foo == { extensible = "foo"; };
+      assert config.merged.extensi-bar == { extensible = "bar"; };
+      # assert lib.foldl' (a: b: builtins.trace b a) true (lib.attrNames config.docs);
```
I've removed it. It doesn't test anything that isn't tested on the next line, and we don't need to print those things.

roberth committed:
lib/tests/modules/types-attrTag.nix: Test other option doc attrs

roberth committed:
lib.types.attrTag: Fix declarationPositions after merge

roberth commented on lib/types.nix:
```diff
@@ -614,6 +620,102 @@ rec {
+                    // {
+                      # mergeOptionDecls is not idempotent in these attrs:
+                      declarations = a.tags.${tagName}.declarations ++ bOpt.declarations;
+                      declarationPositions = a.tags.${tagName}.declarations ++ bOpt.declarations;
```
Fixed and red-greened.

roberth committed:
types.attrTag: Remove unnecessary definitions override

Committing because tests pass.

Co-authored-by: Silvan Mosberger <github@infinisil.com>

infinisil reviewed (approved):
Looking good, let's merge!

The problems I mentioned in https://github.com/NixOS/nixpkgs/pull/284551#pullrequestreview-1961816244 are still not addressed, but it doesn't block this PR :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/home-manager-install-as-module-in-flake-users-option-and-home-manager-cli-not-available/45567/2
